docs/proposals/20191106-machine-load-balancer.md

```diff
@@ -0,0 +1,133 @@
+
+* Introduce one or more extension points to enable integrating external load balancers.
+* Allow the reuse of a Machine Load Balancer across CAPI providers, ex. AWS Elastic Load Balancers (ELB), F5, Netscaler, etc.
+* Represent the status of load balancer in resource status e.g. 4 of 5 members healthy.
+
+#### Non-Goals
+
+* Providing load balancers for non-control plane workloads i.e Service Type Load Balancer
+* Handling other Infrastructure Providers like networking and firewalls, This CAEP may, however, be merged into a larger CAEP when it is ready.
+
+## Summary
+
+The Machine Load Balancer (MLB) provides an abstraction for implementing load balancing across machines while also allowing the resulting  load balancer address to be used by any CRD.
+
+#### Coupling and Dependencies
+
+* The MLB depends on the [template operator](https://docs.google.com/document/d/1dJdnMozsWJ2ccPXWWkZpFVccFFSI7sYaJ6D25sZFlTU/edit#heading=h.vclg2zerysbp) to plumb the load balancer endpoint from the load balancer CRD, to the cluster CRD
+  * The MLB depends on the CAPI Machine CRD'S but it could be extended to other machine like objects as well.
+* CAPI has no dependency on the MLB, it only waits until something sets a `controlPlaneEndpoint`
+
+### Workflow
+
+
+![Figure 1](./images/machine-load-balancer/workflow.png)
+
+
+
+**1 - 4:** Controllers watch the resources they responsible for
+
+**5 - 7:** User or system creates all CRD's *concurrently*
+
+**8 - 9:**  Infra provider spins up infrastructure and then waits until the `spec.controlPlaneEndpoint` is set
+
+**10 -11:** The MLB provisions the load balancer and updates `status.endpoint`
+
+**12:** The template operator picks up that `status.endpoint` has been updated and copies the value to `spec.controlPlaneEndpoint`
+
+**13+** CAPI infra providers pick up that the `controlPlaneEndpoint` has been set and continue provisioning
+
+##   Types
+
+```go
+type MachineLoadBalancer struct {
+  ObjectMeta
+  TypeMeta
+  Spec MachineLoadBalancerSpec
+  Status MachineLoadBalancerStatus
+}
+type MachineLoadBalancerSpec struct {
+  Port MachineLoadBalancerPort
+  // MachineSelector provides a mechanism to select multiple machines based on their labels        // allowing for load balancing across clusters and different network topologies.
+  MachineSelector Selector
+  // ProviderRef is a load balancer specific CRD used for configuring the underlying load                       // balancer and/or providing access credentials
+  ProviderRef corev1.ObjectReference
+}
+
+type MachineProtocol string
+const (
+  TcpMachineProtocol MachineProtocol = "tcp"
+  TcpMachineProtocol MachineProtocol = "udp"
+)
+
+type MachineLoadBalancerPort struct {
+  Port int32
+  MachinePort int32
+  Protocol MachineProtocol
+}
+
+type MachineLoadBalancerStatus struct {
+  // Endpoint is the DNS/IP endpoint for the load balancer (includes scheme and port)
+  Endpoint string
+  // Status provides the health of the load balancer
+  Status string
```

Comment:
Will status as a string field be able to archive the goal of 'Represent the status of load balancer in resource status e.g. 4 of 5 members healthy'?